include/rmx/rmx.hpp

```diff
@@ -0,0 +1,85 @@
+#pragma once
+#include <functional>
+#include <mutex>
+#include <optional>
+#include <type_traits>
+
+namespace rmx {
+
+//! An RAII-style guard wrapping a reference to some type protected by a mutex
+//!
+//! Acquire a `MutexGuard` by locking a `Mutex`.
+template<typename ValueT, typename MutexImplT>
+class MutexGuard
+{
+  public:
+    explicit MutexGuard(ValueT& value_ref, std::unique_lock<MutexImplT>&& lock) noexcept :
+        m_lock(std::move(lock)), m_ref(value_ref)
+    {
+    }
+
+    //! Access the underlying value by reference
+    //!
+    //! @warning It is incorrect to store the reference returned by this operator.
+    [[nodiscard]] ValueT& operator*() noexcept { return m_ref; }
+    [[nodiscard]] const ValueT& operator*() const noexcept { return m_ref; }
+
+    //! Access the underlying value by pointer
+    //!
+    //! @warning It is incorrect to store the pointer returned by this operator.
+    [[nodiscard]] ValueT* operator->() noexcept { return &m_ref.get(); }
+    [[nodiscard]] const ValueT* operator->() const noexcept { return &m_ref.get(); }
+
+    //! Provide access to the underlying std::unique_lock to facilitate use with
+    //! std::condition_variable::wait()
+    //!
+    //! @warning It is incorrect to access the wrapped value if the inner lock has been manually
+    //! unlocked. Don't do that.
+    [[nodiscard]] std::unique_lock<MutexImplT>& inner() noexcept { return m_lock; }
+
+  private:
+    std::unique_lock<MutexImplT> m_lock;
+    std::reference_wrapper<ValueT> m_ref;
+};
+
+//! A Rust-inspired mutex that wraps some other type.
+template<typename ValueT, typename MutexImplT = std::mutex>
+class Mutex
+{
+  public:
+    //! Take ownership of an existing ValueT
+    explicit Mutex(ValueT&& value) noexcept : m_value(std::move(value)) {}
+
+    //! Construct a new ValueT from the given args, includes default constructor
+    template<typename... ArgsT,
+             typename std::enable_if_t<std::is_constructible_v<ValueT, ArgsT...>, bool> = true>
+    explicit Mutex(ArgsT&&... args) : m_value(std::forward<ArgsT>(args)...)
```

Comment:
This isn't enough to construct a POD struct; I had to give it either a default or value constructor (see the tests).

Reply:
that makes sense - POD types don't have a constructor per-se, they just can be initialized by an initializer list. I'm not sure if there's an equivalent template forwarding rule that you could use to do that here. Maybe you could do `m_value{std::forward<ArgsT>(args)l...}`? I'm not sure if that's equivalent or not to what you currently have. It _might_ just solve it

Reply:
Yeah, is_constructible doesn't work for POD types, and adding an overload using is_pod causes other problems (ambiguous overload when you have a type that's both POD and constructible)